src/core/qgsdiagramrenderer.cpp

```diff
@@ -804,6 +857,273 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+      s.backgroundColor = properties.valueAsColor( QgsDiagramLayerSettings::Property::BackgroundColor, c.expressionContext(), s.backgroundColor );
+      c.expressionContext().setOriginalValueVariable( QgsColorUtils::colorToString( s.penColor ) );
+      s.penColor = properties.valueAsColor( QgsDiagramLayerSettings::Property::StrokeColor, c.expressionContext(), s.penColor );
+      c.expressionContext().setOriginalValueVariable( s.penWidth );
+      s.penWidth = properties.valueAsDouble( QgsDiagramLayerSettings::Property::StrokeWidth, c.expressionContext(), s.penWidth );
+      c.expressionContext().setOriginalValueVariable( s.rotationOffset );
+      s.rotationOffset = properties.valueAsDouble( QgsDiagramLayerSettings::Property::StartAngle, c.expressionContext(), s.rotationOffset );
+    }
+
+    QgsPaintEffect *effect = s.paintEffect();
+    std::unique_ptr< QgsEffectPainter > effectPainter;
+    if ( effect && effect->enabled() )
+    {
+      effectPainter = std::make_unique< QgsEffectPainter >( c, effect );
+    }
+
+    stackedRenderer->diagram()->renderDiagram( feature, c, s, newPos );
+    QgsStackedDiagram *stackedDiagram = dynamic_cast< QgsStackedDiagram *>( mDiagram.get() );
+    stackedDiagram->subDiagramPosition( newPos, c, mSettings, s );
+  }
+}
+
+bool QgsStackedDiagramRenderer::diagramSettings( const QgsFeature &feature, const QgsRenderContext &c, QgsDiagramSettings &s ) const
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  Q_UNUSED( s )
+  return false;
+}
+
+QSizeF QgsStackedDiagramRenderer::diagramSize( const QgsFeature &feature, const QgsRenderContext &c ) const
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  return QSizeF( 0, 0 );
+}
+
+QList<QgsDiagramSettings> QgsStackedDiagramRenderer::diagramSettings() const
+{
+  QList<QgsDiagramSettings> settingsList;
+  settingsList.push_back( mSettings );
+  return settingsList;
+}
+
+QList<QString> QgsStackedDiagramRenderer::diagramAttributes() const
+{
+  return mSettings.categoryAttributes;
+}
+
+QList< QgsLayerTreeModelLegendNode * > QgsStackedDiagramRenderer::legendItems( QgsLayerTreeLayer *nodeLayer ) const
+{
+  QList< QgsLayerTreeModelLegendNode * > nodes;
+  for ( const auto &renderer : std::as_const( mDiagramRenderers ) )
+  {
+    nodes << renderer->legendItems( nodeLayer );
+  }
+
+  return nodes;
+}
+
+QList< QgsDiagramRenderer * > QgsStackedDiagramRenderer::renderers( bool sortByDiagramMode ) const
+{
+  QList< QgsDiagramRenderer * > renderers = mDiagramRenderers;
+
+  if ( sortByDiagramMode && mSettings.stackedDiagramMode == QgsDiagramSettings::Vertical )
+  {
+    // We draw vertical diagrams backwards, so
+    // we return the subrenderers in reverse order
+    std::reverse( renderers.begin(), renderers.end() );
+  }
+  return renderers;
+}
+
+void QgsStackedDiagramRenderer::addRenderer( QgsDiagramRenderer *renderer )
+{
+  if ( renderer )
+  {
+    mDiagramRenderers.append( renderer );
```

Comment:
@gacarrillor What is supposed to happen with ownership here? Right now mDiagramRenderers is never cleaned up, so these are all leaked. There's also no SIP_TRANSFER annotation on the function declaration, so you'll get a crash if you call this method from python.

There's also a (very bad) ownership issue with clone() here, because that calls a default copy constructor which just directly copies mDiagramRenderers to the clone, even though all the pointers in that container belong to the original instance. 

Can you please address this ASAP? We are closing in on release and this API is broken in its current state...

Reply:
I'll have a look at it, thanks a lot for pointing this out!

Reply:
Addressed in https://github.com/qgis/QGIS/pull/59123